web/src/main/java/org/springframework/security/web/server/authentication/AnonymousAuthenticationWebFilter.java

```diff
@@ -0,0 +1,97 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.web.server.authentication;
+
+import org.springframework.security.authentication.AnonymousAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.AuthorityUtils;
+import org.springframework.security.core.context.ReactiveSecurityContextHolder;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextImpl;
+import org.springframework.util.Assert;
+import org.springframework.web.server.ServerWebExchange;
+import org.springframework.web.server.WebFilter;
+import org.springframework.web.server.WebFilterChain;
+import reactor.core.publisher.Mono;
+import reactor.util.context.Context;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Detects if there is no {@code Authentication} object in the
+ * {@code SecurityContextHolder}, and populates it with one if needed.
+ *
+ * @author Ankur Pathak
+ * @since 5.2.0
+ */
+public class AnonymousAuthenticationWebFilter implements WebFilter {
+	// ~ Instance fields
+	// ================================================================================================
+
+	private String key;
+	private Object principal;
+	private List<GrantedAuthority> authorities;
+
+	/**
+	 * Creates a filter with a principal named "anonymousUser" and the single authority
+	 * "ROLE_ANONYMOUS".
+	 *
+	 * @param key the key to identify tokens created by this filter
+	 */
+	public AnonymousAuthenticationWebFilter(String key) {
+		this(key, "anonymousUser", AuthorityUtils.createAuthorityList("ROLE_ANONYMOUS"));
+	}
+
+	/**
+	 * @param key         key the key to identify tokens created by this filter
+	 * @param principal   the principal which will be used to represent anonymous users
+	 * @param authorities the authority list for anonymous users
+	 */
+	public AnonymousAuthenticationWebFilter(String key, Object principal,
+											List<GrantedAuthority> authorities) {
+		Assert.hasLength(key, "key cannot be null or empty");
+		Assert.notNull(principal, "Anonymous authentication principal must be set");
+		Assert.notNull(authorities, "Anonymous authorities must be set");
+		this.key = key;
+		this.principal = principal;
+		this.authorities = authorities;
+	}
+
+
+	@Override
+	public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
+		Context context = ReactiveSecurityContextHolder.getContext()
+				.filter(Objects::isNull)
+				.switchIfEmpty(Mono.fromSupplier(() -> {
+					SecurityContext securityContext = new SecurityContextImpl();
+					securityContext.setAuthentication(createAuthentication(exchange));
+					return securityContext;
+				}))
+				.as(ReactiveSecurityContextHolder::withSecurityContext);
+		return Mono.subscriberContext()
```

Comment:
We should only set the subscriberContext if it changes

Reply:
Can you please push the code and point me to the tests that are failing?

Reply:
1. I have pushed my changes.
2. Here is a list:
```
org.springframework.security.config.web.server.RequestCacheTests > requestCacheNoOp FAILED
    org.junit.ComparisonFailure at RequestCacheTests.java:96

org.springframework.security.config.web.server.RequestCacheTests > defaultFormLoginRequestCache FAILED
    org.junit.ComparisonFailure at RequestCacheTests.java:64

org.springframework.security.config.web.server.OAuth2LoginTests > defaultLoginPageWithMultipleClientRegistrationsThenLinks FAILED
    org.junit.ComparisonFailure at OAuth2LoginTests.java:89

org.springframework.security.config.web.server.OAuth2LoginTests > defaultLoginPageWithSingleClientRegistrationThenRedirect FAILED
    java.lang.AssertionError at OAuth2LoginTests.java:123

org.springframework.security.config.web.server.FormLoginTests > defaultLoginPage FAILED
    org.junit.ComparisonFailure at FormLoginTests.java:69

org.springframework.security.config.web.server.FormLoginTests > customLoginPage FAILED
    org.junit.ComparisonFailure at FormLoginTests.java:115

org.springframework.security.config.web.server.LogoutSpecTests > customLogout FAILED
    org.junit.ComparisonFailure at LogoutSpecTests.java:99

org.springframework.security.config.web.server.LogoutSpecTests > defaultLogout FAILED
    org.junit.ComparisonFailure at LogoutSpecTests.java:54

org.springframework.security.config.web.server.AuthorizeExchangeSpecTests > antMatchersWhenMethodAndPatternsThenDiscriminatesByMethod FAILED
    java.lang.AssertionError at AuthorizeExchangeSpecTests.java:55
        Caused by: java.lang.AssertionError at AuthorizeExchangeSpecTests.java:55

org.springframework.security.config.web.server.AuthorizeExchangeSpecTests > antMatchersWhenPatternsThenAnyMethod FAILED
    java.lang.AssertionError at AuthorizeExchangeSpecTests.java:77
        Caused by: java.lang.AssertionError at AuthorizeExchangeSpecTests.java:77

org.springframework.security.config.web.server.ExceptionHandlingSpecTests > defaultAuthenticationEntryPoint FAILED
    java.lang.AssertionError at ExceptionHandlingSpecTests.java:55
        Caused by: java.lang.AssertionError at ExceptionHandlingSpecTests.java:55

org.springframework.security.config.web.server.ExceptionHandlingSpecTests > customAuthenticationEntryPoint FAILED
    java.lang.AssertionError at ExceptionHandlingSpecTests.java:79
        Caused by: java.lang.AssertionError at ExceptionHandlingSpecTests.java:79

org.springframework.security.config.web.server.ServerHttpSecurityTests > basicWhenNoCredentialsThenUnauthorized FAILED
    java.lang.AssertionError at ServerHttpSecurityTests.java:138
        Caused by: java.lang.AssertionError at ServerHttpSecurityTests.java:138

org.springframework.security.config.web.server.ServerHttpSecurityTests > defaults FAILED
    java.lang.IllegalStateException at ServerHttpSecurityTests.java:96

org.springframework.security.config.web.server.OAuth2ResourceServerSpecTests > getWhenCustomBearerTokenEntryPointThenResponds FAILED
    java.lang.AssertionError at OAuth2ResourceServerSpecTests.java:260
        Caused by: java.lang.AssertionError at OAuth2ResourceServerSpecTests.java:260

org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurityTests > defaultMediaAllThenUnAuthorized FAILED
    java.lang.AssertionError at EnableWebFluxSecurityTests.java:113
        Caused by: java.lang.AssertionError at EnableWebFluxSecurityTests.java:113

org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurityTests > defaultRequiresAuthentication FAILED
    java.lang.AssertionError at EnableWebFluxSecurityTests.java:96
        Caused by: java.lang.AssertionError at EnableWebFluxSecurityTests.java:96

org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurityTests > multiWorks FAILED
    java.lang.AssertionError at EnableWebFluxSecurityTests.java:298
        Caused by: java.lang.AssertionError at EnableWebFluxSecurityTests.java:298
```